crates/executor/src/estimate.rs

```diff
@@ -13,54 +19,359 @@ pub fn estimate(
+                transaction_hash = %transaction_hash(&tx),
+                transaction_index = %tx_index
+            )
+            .entered();
 
-        let fee_type = super::transaction::fee_type(&transaction);
-        let gas_vector_computation_mode =
-            super::transaction::gas_vector_computation_mode(&transaction);
-        let minimal_l1_gas_amount_vector = match &transaction {
+            let gas_vector_computation_mode = super::transaction::gas_vector_computation_mode(&tx);
+            let tx_info = match gas_vector_computation_mode {
+                GasVectorComputationMode::NoL2Gas => {
+                    execute_transaction(&tx, tx_index, &mut state, &block_context)
+                }
+                GasVectorComputationMode::All => find_l2_gas_limit_and_execute_transaction(
+                    &mut tx,
+                    tx_index,
+                    &mut state,
+                    &block_context,
+                ),
+            }?;
+
+            tracing::trace!(
+                actual_fee = %tx_info.receipt.fee.0,
+                actual_resources = ?tx_info.receipt.resources,
+                "Transaction estimation finished"
+            );
+
+            Ok(FeeEstimate::from_tx_and_tx_info(
+                &tx,
+                &tx_info,
+                &gas_vector_computation_mode,
+                &block_context,
+            ))
+        })
+        .collect()
+}
+
+/// The margin for the binary search for the minimal L2 gas limit.
+const L2_GAS_SEARCH_MARGIN: GasAmount = GasAmount(100);
+
+/// Searches for the minimal L2 gas limit (within a certain margin) that allows
+/// the transaction to execute without running out of L2 gas. Uses this limit to
+/// execute the transaction. Exceeding the user provided L2 gas limit is not
+/// allowed.
+///
+/// This is needed because Starknet 0.13.4 introduced runtime L2 gas accounting
+/// which could lead to transactions being reverted because of insufficient L2
+/// gas, even though the limit was set to the L2 gas cost of the transaction
+/// (because the worst-case path gas requirements are larger than the actual
+/// cost).
+fn find_l2_gas_limit_and_execute_transaction<S>(
+    tx: &mut Transaction,
+    tx_index: usize,
+    state: &mut S,
+    block_context: &blockifier::context::BlockContext,
+) -> Result<TransactionExecutionInfo, TransactionExecutionError>
+where
+    S: UpdatableState,
+{
+    let initial_l2_gas_limit = get_l2_gas_limit(tx);
+
+    // Start with MAX gas limit to get the consumed L2 gas.
+    update_l2_gas_limit(tx, GasAmount::MAX);
+    let tx_info = match simulate_transaction(tx, tx_index, state, block_context) {
+        Ok(tx_info) => tx_info,
+        Err(TransactionSimulationError::ExecutionError(error)) => {
+            return Err(error);
+        }
+        Err(TransactionSimulationError::OutOfGas) => {
+            unreachable!("Transaction should not run out of gas with MAX gas limit")
```

Comment:
I wonder if it's possible to construct a transaction so expensive it would run out even of `GasAmount::MAX`...

Reply:
Plus, that seems to me a trivial DoS vector. I'm sure there's a per-block gas limit -- we should probably use that as the maximum?

Reply:
Plus, that seems to be a trivial DoS vector -- I'm sure there's a per-block gas limit, we should probably just use that.